<a href="https://colab.research.google.com/github/Sanket-S-Kale/Prompt-Engineering-Programming/blob/main/Copy_of_M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [20]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipykernel_5254/1930504112.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   250.000000  250.000000  250.000000  250.000000  2.500000e+02
mean    170.351416  171.642984  168.804713  170.063142  5.922355e+07
std      17.324150   17.256061   17.522302   17.519914  1.778332e+07
min     123.046814  125.753411  122.210212  124.021217  2.404830e+07
25%     159.797470  160.068566  158.405160  159.247956  4.781208e+07
50%     173.442314  174.974387  171.826260  173.214974  5.511085e+07
75%     185.253357  185.885956  182.783124  184.392075  6.574292e+07
max     196.073090  197.567559  194.974504  195.984019  1.543573e+08


In [21]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Summarize key statistics
print(data.describe())


/tmp/ipykernel_5254/1631210751.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   250.000000  250.000000  250.000000  250.000000  2.500000e+02
mean    170.351416  171.642984  168.804713  170.063142  5.922355e+07
std      17.324150   17.256061   17.522302   17.519914  1.778332e+07
min     123.046814  125.753411  122.210212  124.021217  2.404830e+07
25%     159.797470  160.068566  158.405160  159.247956  4.781208e+07
50%     173.442314  174.974387  171.826260  173.214974  5.511085e+07
75%     185.253357  185.885956  182.783124  184.392075  6.574292e+07
max     196.073090  197.567559  194.974504  195.984019  1.543573e+08


In [22]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipykernel_5254/1212492746.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   250.000000  250.000000  250.000000  250.000000  2.500000e+02
mean    170.351416  171.642984  168.804713  170.063142  5.922355e+07
std      17.324150   17.256061   17.522302   17.519914  1.778332e+07
min     123.046814  125.753411  122.210212  124.021217  2.404830e+07
25%     159.797470  160.068566  158.405160  159.247956  4.781208e+07
50%     173.442314  174.974387  171.826260  173.214974  5.511085e+07
75%     185.253357  185.885956  182.783124  184.392075  6.574292e+07
max     196.073090  197.567559  194.974504  195.984019  1.543573e+08


## Step 2: Feature Engineering

I will now create additional features based on the historical stock data, including moving averages and a target variable for classification.

## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

## Step 3.1: Model Building - Linear Regression for Price Prediction

I will train a regression model to predict tomorrow's stock price using the specified features.

In [24]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
import ta # Ensure 'ta' library is installed for technical indicators
import pandas as pd

# Flatten MultiIndex columns if they exist (common with yfinance output)
if isinstance(data.columns, pd.MultiIndex):
    data.columns = data.columns.get_level_values(0)

# Recalculate features to ensure they are present in the 'data' DataFrame
# Calculate Moving Averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate RSI
data['RSI'] = ta.momentum.RSIIndicator(data['Close'], window=14).rsi()

# Calculate MACD
macd = ta.trend.MACD(data['Close'])
data['MACD'] = macd.macd()
data['MACD_Signal'] = macd.macd_signal()

# Calculate ATR (Volatility)
data['ATR'] = ta.volatility.AverageTrueRange(data['High'], data['Low'], data['Close'], window=14).average_true_range()

# Add Time-based features
data['Year'] = data.index.year
data['Month'] = data.index.month
data['Day'] = data.index.day
data['DayOfWeek'] = data.index.dayofweek

# Create the target variable for regression: tomorrow's close price
data['Tomorrow_Close_Regression'] = data['Close'].shift(-1)

# Drop rows with NaN values that result from feature calculations and the shifting for the target variable
data_regression = data.dropna().copy()

# Define features (X) and target (y) for regression
feature_cols = ['Close', 'High', 'Low', 'Open', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'RSI', 'MACD', 'MACD_Signal', 'ATR', 'Year', 'Month', 'Day', 'DayOfWeek']

# Filter out features that might not be present if the 'ta' library wasn't installed or data was too short
feature_cols = [col for col in feature_cols if col in data_regression.columns]

X = data_regression[feature_cols]
y = data_regression['Tomorrow_Close_Regression']

# Split data into training and testing sets (70/30 ratio)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=False)

print(f"Training set size: {len(X_train)} samples")
print(f"Testing set size: {len(X_test)} samples")

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"\nMean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R²): {r2:.3f}")

if r2 < 0:
    print("\nThese metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy.")
elif r2 < 0.5:
    print("\nThe R-squared value indicates how much of the variance in the dependent variable can be explained by the independent variables. A higher R-squared value indicates a better fit for the model. However, an R-squared below 0.5 might suggest that the model only moderately explains the variance, and further improvements could be beneficial.")
else:
    print("\nThe R-squared value indicates how much of the variance in the dependent variable can be explained by the independent variables. A higher R-squared value indicates a better fit for the model.")

Training set size: 140 samples
Testing set size: 60 samples

Mean Squared Error (MSE): 8.11
R-squared (R²): 0.896

The R-squared value indicates how much of the variance in the dependent variable can be explained by the independent variables. A higher R-squared value indicates a better fit for the model.


## Step 3.2: Model Building - Random Forest Classification

I will now create a binary target variable and train a Random Forest model for classification.

In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import pandas as pd

# Ensure MultiIndex columns are flattened if they exist
if isinstance(data.columns, pd.MultiIndex):
    data.columns = data.columns.get_level_values(0)

# Create the binary target variable for classification
# 1 if tomorrow's close price is higher than today's, 0 otherwise
data['Tomorrow_Close_Classification'] = data['Close'].shift(-1)
data['target_classification'] = (data['Tomorrow_Close_Classification'] > data['Close']).astype(int)

# Drop rows with NaN values that result from the shifting for the target variable
data_classification = data.dropna().copy()

# Define features (X) and target (y) for classification
feature_cols = ['Close', 'High', 'Low', 'Open', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'RSI', 'MACD', 'MACD_Signal', 'ATR', 'Year', 'Month', 'Day', 'DayOfWeek']

# Filter out features that might not be present
feature_cols = [col for col in feature_cols if col in data_classification.columns]

X_clf = data_classification[feature_cols]
y_clf = data_classification['target_classification']

# Split data into training and testing sets (70/30 ratio), maintaining the chronological order
X_train_clf, X_test_clf, y_train_clf, y_test_clf = train_test_split(X_clf, y_clf, test_size=0.3, shuffle=False)

print(f"Training set size (Classification): {len(X_train_clf)} samples")
print(f"Testing set size (Classification): {len(X_test_clf)} samples")

# Initialize and train the Random Forest Classifier model
clf_model = RandomForestClassifier(n_estimators=100, random_state=42)
clf_model.fit(X_train_clf, y_train_clf)

# Make predictions on the test set
y_pred_clf = clf_model.predict(X_test_clf)

# Evaluate the model
accuracy = accuracy_score(y_test_clf, y_pred_clf)
conf_matrix = confusion_matrix(y_test_clf, y_pred_clf)
class_report = classification_report(y_test_clf, y_pred_clf)

print(f"\nAccuracy: {accuracy:.2f}")
print(f"\nConfusion Matrix:\n{conf_matrix}")
print(f"\nClassification Report:\n{class_report}")

# Report Feature Importance
feature_importance = pd.DataFrame({'Feature': feature_cols, 'Importance': clf_model.feature_importances_})
feature_importance = feature_importance.sort_values(by='Importance', ascending=False)

print("\nFeature Importance:")
display(feature_importance)

Training set size (Classification): 140 samples
Testing set size (Classification): 60 samples

Accuracy: 0.52

Confusion Matrix:
[[16 11]
 [18 15]]

Classification Report:
              precision    recall  f1-score   support

           0       0.47      0.59      0.52        27
           1       0.58      0.45      0.51        33

    accuracy                           0.52        60
   macro avg       0.52      0.52      0.52        60
weighted avg       0.53      0.52      0.52        60


Feature Importance:


,Feature,Importance
14,Day,0.120294
15,DayOfWeek,0.091763
9,MACD,0.088824
8,RSI,0.084412
10,MACD_Signal,0.068789
11,ATR,0.066367
0,Close,0.062988
4,Volume,0.060686
7,MA_50,0.059064
5,MA_5,0.059000


## Step 3.3: Hyperparameter Tuning for Random Forest Classification

I will use `GridSearchCV` to find the optimal hyperparameters for the Random Forest Classifier to improve its performance.

In [26]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Define the parameter grid to search
param_grid = {
    'n_estimators': [50, 100, 150],
    'max_features': ['sqrt', 'log2'],
    'max_depth': [5, 10, None],
    'criterion': ['gini', 'entropy']
}

# Initialize a Random Forest Classifier
rf_base = RandomForestClassifier(random_state=42)

# Fit GridSearchCV to the training data
grid_search = GridSearchCV(estimator=rf_base, param_grid=param_grid,
                           cv=5, n_jobs=-1, verbose=2, scoring='accuracy')

# Fit GridSearchCV to the training data
grid_search.fit(X_train_clf, y_train_clf)

print(f"Best parameters found: {grid_search.best_params_}")
print(f"Best cross-validation accuracy: {grid_search.best_score_:.2f}")

# Get the best model
best_rf_model = grid_search.best_estimator_

# Make predictions on the test set with the best model
y_pred_tuned = best_rf_model.predict(X_test_clf)

# Evaluate the tuned model
accuracy_tuned = accuracy_score(y_test_clf, y_pred_tuned)
conf_matrix_tuned = confusion_matrix(y_test_clf, y_pred_tuned)
class_report_tuned = classification_report(y_test_clf, y_pred_tuned)

print(f"\nOptimized Model Accuracy: {accuracy_tuned:.2f}")
print(f"\nOptimized Model Confusion Matrix:\n{conf_matrix_tuned}")
print(f"\nOptimized Model Classification Report:\n{class_report_tuned}")

# Report Feature Importance from the best model
feature_cols = ['Close', 'High', 'Low', 'Open', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'RSI', 'MACD', 'MACD_Signal', 'ATR', 'Year', 'Month', 'Day', 'DayOfWeek'] # Update feature_cols for display
feature_importance_tuned = pd.DataFrame({'Feature': feature_cols, 'Importance': best_rf_model.feature_importances_})
feature_importance_tuned = feature_importance_tuned.sort_values(by='Importance', ascending=False)

print("\nOptimized Feature Importance:")
display(feature_importance_tuned)

Fitting 5 folds for each of 36 candidates, totalling 180 fits
Best parameters found: {'criterion': 'gini', 'max_depth': None, 'max_features': 'sqrt', 'n_estimators': 50}
Best cross-validation accuracy: 0.48

Optimized Model Accuracy: 0.55

Optimized Model Confusion Matrix:
[[17 10]
 [17 16]]

Optimized Model Classification Report:
              precision    recall  f1-score   support

           0       0.50      0.63      0.56        27
           1       0.62      0.48      0.54        33

    accuracy                           0.55        60
   macro avg       0.56      0.56      0.55        60
weighted avg       0.56      0.55      0.55        60


Optimized Feature Importance:


,Feature,Importance
14,Day,0.124484
8,RSI,0.089402
15,DayOfWeek,0.083211
9,MACD,0.082289
10,MACD_Signal,0.065806
11,ATR,0.063369
5,MA_5,0.063231
6,MA_10,0.061909
0,Close,0.061665
2,Low,0.061585


## Step 4: Model Evaluation & Prediction (Optimized Random Forest)

In [28]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import pandas as pd

# Make predictions on the test set with the best tuned model
y_pred_tuned_final = best_rf_model.predict(X_test_clf)

# Evaluate the tuned model
accuracy_tuned_final = accuracy_score(y_test_clf, y_pred_tuned_final)
conf_matrix_tuned_final = confusion_matrix(y_test_clf, y_pred_tuned_final)
class_report_tuned_final = classification_report(y_test_clf, y_pred_tuned_final)

print(f"\nOptimized Model Accuracy for Prediction Period: {accuracy_tuned_final:.2f}")
print(f"\nOptimized Model Confusion Matrix for Prediction Period:\n{conf_matrix_tuned_final}")
print(f"\nOptimized Model Classification Report for Prediction Period:\n{class_report_tuned_final}")


Optimized Model Accuracy for Prediction Period: 0.55

Optimized Model Confusion Matrix for Prediction Period:
[[17 10]
 [17 16]]

Optimized Model Classification Report for Prediction Period:
              precision    recall  f1-score   support

           0       0.50      0.63      0.56        27
           1       0.62      0.48      0.54        33

    accuracy                           0.55        60
   macro avg       0.56      0.56      0.55        60
weighted avg       0.56      0.55      0.55        60



# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.

